program/src/processor.rs

```diff
@@ -355,7 +356,113 @@ fn process_initialize_holder_rewards(
+        if !holder_rewards_pool_info
+            .key
+            .eq(&get_holder_rewards_pool_address(mint_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsPoolAddress.into());
+        }
+
+        let holder_rewards_pool_data = holder_rewards_pool_info.try_borrow_data()?;
+        let holder_rewards_pool_state =
+            bytemuck::try_from_bytes::<HolderRewardsPool>(&holder_rewards_pool_data)
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+
+        holder_rewards_pool_state.total_rewards
+    };
+
+    let (last_seen_total_rewards, unharvested_rewards) = {
+        // Ensure the holder rewards account is owned by the Paladin Rewards
+        // program.
+        if !holder_rewards_info.owner.eq(program_id) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the provided holder rewards address is the correct address
+        // derived from the token account.
+        if !holder_rewards_info
+            .key
+            .eq(&get_holder_rewards_address(token_account_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsAddress.into());
+        }
+
+        let mut holder_rewards_data = holder_rewards_info.try_borrow_mut_data()?;
+        let holder_rewards_state =
+            bytemuck::try_from_bytes_mut::<HolderRewards>(&mut holder_rewards_data)
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+
+        // Update the holder rewards state with the new "last seen" total
+        // rewards and zero out the unharvested rewards.
+        (
+            std::mem::replace(
+                &mut holder_rewards_state.last_seen_total_rewards,
+                current_total_rewards,
+            ),
+            std::mem::take(&mut holder_rewards_state.unharvested_rewards),
+        )
+    };
+
+    // Calculate unharvested rewards for the token account.
+    // Since the holder rewards account may already have unharvested rewards,
+    // calculate the share of rewards that have not been seen by the holder
+    // rewards account.
+    let unseen_total_rewards = current_total_rewards
+        .checked_sub(last_seen_total_rewards)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    let unseen_unharvested_rewards =
+        calculate_reward_share(token_supply, token_account_balance, unseen_total_rewards)?;
+    let rewards_to_harvest = unharvested_rewards
+        .checked_add(unseen_unharvested_rewards)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    // Move the amount from the holder rewards pool to the token account.
+    let new_holder_rewards_pool_lamports = holder_rewards_pool_info
+        .lamports()
+        .checked_sub(rewards_to_harvest)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    let new_token_account_lamports = token_account_info
+        .lamports()
+        .checked_add(rewards_to_harvest)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
```

Comment:
Based on our conversations, there are situations where we can overdraw the pool account. In those cases, we should take `min(rewards_to_harvest, lamports - rent_exemption`.

Could you add a test case for the overdrawing behavior?

Reply:
Added to program and tests!